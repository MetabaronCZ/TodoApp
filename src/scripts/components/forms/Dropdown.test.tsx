import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';

import { Dropdown, DropdownItem } from 'components/forms/Dropdown';
import { withMockedProviders } from 'test/component';

const testItems: DropdownItem<string>[] = [
  { id: '1', title: 'Item 1', value: '1' },
  { id: '2', title: 'Item 2', value: '2' },
  { id: '3', title: 'Item 3', value: '3' },
];

describe('components/forms/Dropdown', () => {
  it('should render correctly', () => {
    const { container } = render(
      withMockedProviders(
        <Dropdown items={testItems} value="1" onSelect={() => null} />,
      ),
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelector('button')).toBeInTheDocument();
  });

  it('should be open-able', () => {
    const tree = render(
      withMockedProviders(
        <Dropdown items={testItems} value="1" onSelect={() => null} />,
      ),
    );
    expect(tree.queryByRole('list')).not.toBeInTheDocument();

    fireEvent.click(tree.getByRole('button'));
    expect(tree.container).toMatchSnapshot();
    expect(tree.getByRole('list')).toBeInTheDocument();

    const items = tree.container.querySelectorAll('ul button');
    expect(items[0].textContent).toEqual('Item 1');
    expect(items[1].textContent).toEqual('Item 2');
    expect(items[2].textContent).toEqual('Item 3');
  });

  it('should be close-able', () => {
    const tree = render(
      withMockedProviders(
        <Dropdown items={testItems} value="1" onSelect={() => null} />,
      ),
    );
    expect(tree.queryByRole('list')).not.toBeInTheDocument();

    // open
    fireEvent.click(tree.getByRole('button'));
    expect(tree.getByRole('list')).toBeInTheDocument();

    // close by clicking the Dropdown button
    fireEvent.click(tree.getAllByRole('button')[0]);
    expect(tree.queryByRole('list')).not.toBeInTheDocument();

    // open
    fireEvent.click(tree.getByRole('button'));
    expect(tree.getByRole('list')).toBeInTheDocument();

    // close by clicking outside Dropdown
    fireEvent.mouseUp(tree.container);
    expect(tree.queryByRole('list')).not.toBeInTheDocument();

    // open
    fireEvent.click(tree.getByRole('button'));
    expect(tree.getByRole('list')).toBeInTheDocument();

    // close by pressing ESC
    fireEvent.keyDown(tree.container, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27,
    });
    expect(tree.queryByRole('list')).not.toBeInTheDocument();
  });

  it('should be disabled', () => {
    const tree = render(
      withMockedProviders(
        <Dropdown items={testItems} value="1" disabled onSelect={() => null} />,
      ),
    );
    expect(tree.queryByRole('list')).not.toBeInTheDocument();

    fireEvent.click(tree.getByRole('button'));
    expect(tree.queryByRole('list')).not.toBeInTheDocument();
  });

  it('should be left-alignable', () => {
    const tree = render(
      withMockedProviders(
        <Dropdown
          items={testItems}
          align="left"
          value="1"
          onSelect={() => null}
        />,
      ),
    );

    fireEvent.click(tree.getByRole('button'));
    expect(tree.getByRole('list')).toBeInTheDocument();
    expect(tree.getByRole('list')).toHaveStyle({ left: 0 });
  });

  it('should be right-alignable', () => {
    const tree = render(
      withMockedProviders(
        <Dropdown
          items={testItems}
          align="right"
          value="1"
          onSelect={() => null}
        />,
      ),
    );

    fireEvent.click(tree.getByRole('button'));
    expect(tree.getByRole('list')).toBeInTheDocument();
    expect(tree.getByRole('list')).toHaveStyle({ right: 0 });
  });

  it('should select items', () => {
    const onSelect = jest.fn();

    const tree = render(
      withMockedProviders(
        <Dropdown items={testItems} value="1" onSelect={onSelect} />,
      ),
    );

    // open Dropdown
    fireEvent.click(tree.getByRole('button'));
    expect(tree.getByRole('list')).toBeInTheDocument();
    expect(onSelect).toBeCalledTimes(0);

    const buttons = tree.getAllByRole('button');
    fireEvent.click(buttons[2]); // click second item

    expect(onSelect).toBeCalledTimes(1);
    expect(onSelect).toBeCalledWith('2');
    expect(tree.queryByRole('list')).not.toBeInTheDocument(); // Dropdown closed
  });
});
