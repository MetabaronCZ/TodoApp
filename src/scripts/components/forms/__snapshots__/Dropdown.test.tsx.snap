// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/forms/Dropdown should be open-able 1`] = `
.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c6 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  background: #f0f0f0;
}

.c6:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c6:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c6:disabled {
  color: #c0c0c0;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  position: relative;
  display: inline-block;
}

.c5 {
  list-style-type: none;
  position: absolute;
  top: 100%;
  left: 0;
  min-width: 160px;
  border: 1px solid #c0c0c0;
  background: #f0f0f0;
  z-index: 10000;
}

.c7 {
  width: 100%;
  text-align: left;
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
      type="button"
    >
      <span
        class="c3"
      >
        Item 1
      </span>
      <i
        class="c4"
      >
        ˄
      </i>
    </button>
    <ul
      class="c5"
    >
      <li
        class=""
      >
        <button
          class="c1 c6 c7"
          type="button"
        >
          <span
            class="c3"
          >
            Item 1
          </span>
        </button>
      </li>
      <li
        class=""
      >
        <button
          class="c1 c6 c7"
          type="button"
        >
          <span
            class="c3"
          >
            Item 2
          </span>
        </button>
      </li>
      <li
        class=""
      >
        <button
          class="c1 c6 c7"
          type="button"
        >
          <span
            class="c3"
          >
            Item 3
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`components/forms/Dropdown should render correctly 1`] = `
.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c2:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  position: relative;
  display: inline-block;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
      type="button"
    >
      <span
        class="c3"
      >
        Item 1
      </span>
      <i
        class="c4"
      >
        ˅
      </i>
    </button>
  </div>
</div>
`;
